### PR TITLE
[ISSUE-718] Pass parameters from helm

### PR DESCRIPTION
**Pull request by rkiselenko**
### Issue dell/csi-baremetal#718

~envVars parameter introduced in order to pass env variables from helm to sidecar container~
args introduced in order to denote entrypoint args from the `values.yaml`.

## PR checklist
- [x] Add link to the issue
- [x] Choose Project
- [x] Choose PR label
- [ ] New unit tests added
- [x] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing

There are two option

## 1. Install from charts placed locally

Change chart file `CHARTS_PATH/values.yml`:

```yml
provisioner:
    image:
      # if you want to use topology feature (multiple PVCs per pod) you should use v1.2.2
      tag: v3.0.0
    args:
      # https://github.com/kubernetes-csi/external-provisioner#command-line-options
      timeout: 30s
      retryIntervalStart: 1s
      retryIntervalMax: 5m
      workerThreads: 100
```

Install chart with `helm`:

```shell
#!/bin/bash
set -e
export CSI_VERSION=1.1.0-526.701e69a
export CSI_OPERATOR_VERSION=1.1.0-101.cc4f9d3
export CHARTS_PATH=/root/csi-baremetal-operator/charts
helm install csi-baremetal-operator "$CHARTS_PATH/csi-baremetal-operator" \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_OPERATOR_VERSION \
        --set image.pullPolicy=IfNotPresent
helm install csi-baremetal "$CHARTS_PATH/csi-baremetal-deployment" \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set scheduler.patcher.enable=true \
        --set driver.drivemgr.type=loopbackmgr \
        --set scheduler.log.level=debug \
        --set nodeController.log.level=debug \
        --set driver.log.level=debug \
        --set scheduler.patcher.readinessTimeout=3 \
        --set driver.drivemgr.deployConfig=true \
```

Here is an example of describe for sidecar container without `values.yml` changes:

```shell
 kubectl describe pods csi-baremetal-controller
  csi-provisioner:
    Container ID:  containerd://9072ab88eccad1bb9207290574d01d486839813c37ba5849f00491e0a988a885
    Image:         csi-provisioner:v3.0.0
    Image ID:      sha256:fe0f921f3c92aaf2167c7c373ae48f2f008c0259b288785432c150e82ab62be8
    Port:          <none>
    Host Port:     <none>
    Args:
      --csi-address=$(ADDRESS)
      --v=5
      --feature-gates=Topology=true
      --extra-create-metadata
      --timeout=30s
      --retry-interval-start=1s
      --retry-interval-max=5m
      --worker-threads=100
```

With changes in `values.yml`:

```shell
 csi-provisioner:
    Container ID:  containerd://5111d5092cc0b9b83e91143e3848fa850115d9568081c965ef351515ab4e9401
    Image:         csi-provisioner:v3.0.0
    Image ID:      sha256:fe0f921f3c92aaf2167c7c373ae48f2f008c0259b288785432c150e82ab62be8
    Port:          <none>
    Host Port:     <none>
    Args:
      --csi-address=$(ADDRESS)
      --extra-create-metadata
      --feature-gates=Topology=true
      --timeout=40s
      --retry-interval-start=1s
      --retry-interval-max=5m
      --worker-threads=300
```

## 2. Install from charts repository

Install:

```shell
#!/bin/bash
set -e
export CSI_VERSION=1.1.0-534.ee880b5
export CSI_OPERATOR_VERSION=1.1.0-104.ada12db
export CHARTS_PATH=/root/csi-baremetal-operator/charts
helm install csi-baremetal-operator "$CHARTS_PATH/csi-baremetal-operator" \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_OPERATOR_VERSION \
        --set image.pullPolicy=IfNotPresent
helm install csi-baremetal "$CHARTS_PATH/csi-baremetal-deployment" \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set scheduler.patcher.enable=true \
        --set driver.drivemgr.type=loopbackmgr \
        --set scheduler.log.level=debug \
        --set nodeController.log.level=debug \
        --set driver.log.level=debug \
        --set scheduler.patcher.readinessTimeout=3 \
        --set driver.drivemgr.deployConfig=true \
        --set driver.provisioner.args.timeout=40s \
        --set driver.provisioner.args.workerThreads=300
```

`kubectl describe name_of_pod`:

```
...
 csi-provisioner:
    Container ID:  containerd://5111d5092cc0b9b83e91143e3848fa850115d9568081c965ef351515ab4e9401
    Image:         csi-provisioner:v3.0.0
    Image ID:      sha256:fe0f921f3c92aaf2167c7c373ae48f2f008c0259b288785432c150e82ab62be8
    Port:          <none>
    Host Port:     <none>
    Args:
      --csi-address=$(ADDRESS)
      --extra-create-metadata
      --feature-gates=Topology=true
      --timeout=40s
      --retry-interval-start=1s
      --retry-interval-max=5m
      --worker-threads=300
...
```

### CODE REVIEW

**Review by yurkov-anton — APPROVED** [review]

**Comment by mishoyama** [comment]
@rkiselenko please provide examples how to use this parameters and fill in testing section

File: charts/csi-baremetal-deployment/values.yaml [hunk 1]
```diff
@@ -79,6 +79,18 @@ driver:
     image:
       # if you want to use topology feature (multiple PVCs per pod) you should use v1.2.2
       tag: v3.0.0
+    envVars:
+      # the following parameters is required https://github.com/kubernetes-csi/external-provisioner#command-line-options
+      - name: ADDRESS
+        value: "/csi/csi.sock"
+      - name: TIMEOUT_DURATION
+        value: 30s
+      - name: RETRY_INTERVAL_START
```
Comment (safronovD) [inline charts/csi-baremetal-deployment/values.yaml:88]: Could we comment envs here and pass the related argument in the controller, if env variable is set only? So we will keep default provisioner variables, if they are changed from version to version. Or is it redundant?

Comment (rkiselenko) [reply charts/csi-baremetal-deployment/values.yaml:88]: That values holds default parameters

Comment (safronovD) [reply charts/csi-baremetal-deployment/values.yaml:88]: I mean to not pass the related arg, if env is excluded. For example:
1. In helm values
```
      - name: TIMEOUT_DURATION
        value: 30s
     #- name: RETRY_INTERVAL_START
     #   value: 1s
```
2. Args
```
["--timeout=$(TIMEOUT_DURATION)",]
```
And no `"--retry-interval-start=$(RETRY_INTERVAL_START)",`

So default provisioner timeout will be used

Comment (rkiselenko) [reply charts/csi-baremetal-deployment/values.yaml:88]: It's kind complex I think. If you omit some args then default values will be used under the hood. I've investigate another way by passing args but found env variables a more flexible solutions.

Comment (safronovD) [reply charts/csi-baremetal-deployment/values.yaml:88]: Got it

**Review by safronovD — APPROVED** [review]

File: charts/csi-baremetal-deployment/values.yaml (hunk 1 above)
Comment (rkiselenko) [reply charts/csi-baremetal-deployment/values.yaml]: You point me to good idea 👍 , I'll refactor my code.

**Comment by mishoyama** [comment]
@rkiselenko it's still not clear how installation or upgrade command will look like when I want to change default parameters for `external-provisioner`...

**Review by mishoyama — COMMENTED** [review]
Please provide `helm install` command example

**Review by yurkov-anton — APPROVED** [review]

**Comment by rkiselenko** [comment]
> Please provide `helm install` command example

Added to testing section

**Comment by mishoyama** [comment]
@rkiselenko could you please make the following `external-provisioner` parameter as a regular helm values?
1. --timeout: driver.provisioner.timeout
2. --retry-interval-start: driver.provisioner.retryIntervalStart
3. --retry-interval-max: driver.provisioner.retryIntervalMax
4. --worker-threads: driver.provisioner.workerThreads

**Review by mishoyama — APPROVED** [review]

File: api/v1/components/sidecar.go [hunk 2]
```diff
@@ -19,7 +19,18 @@ package components
 // Sidecar represent CSI sidecar containers
 type Sidecar struct {
 	Image *Image `json:"image,omitempty"`
+	// Arguments to the entrypoint.
+	// +kubebuilder:validation:Type=object
+	// +kubebuilder:validation:Required
+	Args *args `json:"args,omitempty"`
```
[the remark this answers is not in the document]
Comment (rkiselenko) [reply api/v1/components/sidecar.go:25]: No

[the remark this answers is not in the document]
Comment (safronovD) [reply api/v1/components/sidecar.go:25]: Will it be null?

[the remark this answers is not in the document]
Comment (rkiselenko) [reply api/v1/components/sidecar.go:25]: Yes

**Pull request merged**